Fix passabilities of objects on map

ihhub opened the pull request:
The original code was based on hacks and it wasn't reliable. The current approach uses an algorithm and a set of conditions to determine passability of a tile.

close #1951
close #3750
close #3745
close #1521
close #496
close #2601 
close #2401 
close #2364 
close #3246

ihhub commented:
Hi @LeHerosInconnu, could you please test these changes as the highest priority? The changes only work on new games. To speed up the work you could consider modifying `debug` parameter in configuration file to **65535** giving you an extra powerful hero.

I would like to know that we didn't screw gameplay with these changes as the code is re-written from scratch.

oleg-derevenetz reviewed (commented):
Hi @ihhub Since I'm not very familiar with map objects passability rules, I left a few questions here. Would you mind taking a look?

ihhub commented on src/fheroes2/maps/maps_tiles.cpp:
```diff
@@ -1054,234 +790,99 @@ const fheroes2::Image & Maps::Tiles::GetTileSurface( void ) const
-#endif
-        }
+    const int objId = GetObject( false );
+    const bool isActionObject = MP2::isGroundObject( objId );
```
Excellent question. Even before changes `isGroundObject` covered all objects except `OBJ_COAST` by **obj > 127** condition. `isWaterObject` at the same time markes boats as non-action which leads to a fact that a hero cannot embrak a boat from top tiles. To avoid changing the logic in other places it was decided to use `isGroundObject` method.

oleg-derevenetz replied:
May be rename all this somehow? :) For example, rename `isGroundObject(int)` rename to `isActionObject(int)`, so there will be `isActionObject(int)` (more "generic" one) and `isActionObject(int, bool)` (more "specific" one), and add appropriate comments?

ihhub replied:
Sure. I will do.

ihhub commented on src/fheroes2/maps/maps_tiles.cpp:
```diff
@@ -1054,234 +790,99 @@ const fheroes2::Image & Maps::Tiles::GetTileSurface( void ) const
-    default:
-        break;
+    if ( MP2::isGroundObject( objId ) ) {
+        return MP2::getActionObjectDirection( objId );
```
As explained in the other comment `isGroundObject` covers all action objects.

idshibanov reviewed (approved):
Overall new logic looks good to me, although with such a big PR it's impossible to check every case. Few testing sessions definitely needed.

oleg-derevenetz commented on src/fheroes2/maps/maps_tiles_quantity.cpp:
```diff
@@ -425,6 +425,7 @@ Troop Maps::Tiles::QuantityTroop( void ) const
 
 void Maps::Tiles::QuantityReset( void )
 {
+    // TODO: don't modify first 2 bits of quantity1.
```
AFAIK quantity1 is also used as a highest byte of monster count/resource count on a tile and, for example, is used in `Maps::Tiles::QuantityTroop()` (via `Maps::Tiles::MonsterCount()`) to return a count of monsters defending a dwelling or non-captureable dwelling objects, such as Troll Bridge. Won't this break something?

ihhub replied:
Right now the logic for passability rules checks if a tile is an action object so this code right now won't break anything. However, to prevent any possible issues in the future we have to fix it.

oleg-derevenetz replied:
Hi @ihhub Yes, I noticed that. Passability-related code respects object type. But other code (related to dwellings, etc) is currently unaware that these 2 bits are "special". So if tile contains more than 16383 units (yes, I know...), it may not be reset properly if we will not modify these first 2 bits here.

ihhub replied:
This issue we need to address in a separate pull request. By right once any information is read from mp2 or addon structures we have to extract priority information and use `quantity1` as a normal value. That's why I added these TODO comments.

ihhub replied:
Also right now passability is set at the time of reading a map but in the future we might have a code to remove objects from the map during gameplay which will break passability rules. For now we keep as it is and fix later priority value by storing it in a separate variable without bitwise operations like right now.

oleg-derevenetz replied:
@ihhub

> For now we keep as it is and fix later priority value by storing it in a separate variable without bitwise operations like right now.

Yes, that's what I also wanted to offer.

ihhub commented:
Hi @oleg-derevenetz , I addressed our comments. Could you please take a look again?

LeHerosInconnu commented:
Hello @ihhub,

> Hi @LeHerosInconnu, could you please test these changes as the highest priority? The changes only work on new games. To speed up the work you could consider modifying `debug` parameter in configuration file to **65535** giving you an extra powerful hero.
> 
> I would like to know that we didn't screw gameplay with these changes as the code is re-written from scratch.

It sounds like you are still working on this.
Are there any adjustments that need to be made before the change is in its "final" version?
I wouldn't want to start testing again.
Please tell me when it will be ready for testing.

ihhub commented:
> Hello @ihhub,
> 
> > Hi @LeHerosInconnu, could you please test these changes as the highest priority? The changes only work on new games. To speed up the work you could consider modifying `debug` parameter in configuration file to **65535** giving you an extra powerful hero.
> > I would like to know that we didn't screw gameplay with these changes as the code is re-written from scratch.
> 
> It sounds like you are still working on this.
> Are there any adjustments that need to be made before the change is in its "final" version?
> I wouldn't want to start testing again.
> Please tell me when it will be ready for testing.

Hi @LeHerosInconnu , I was addressing code related issues without touching the game's logic. Also right now all reviewers approved the changes so please proceed with testing once the current commit will be ready to be downloaded (in maximum 20 minutes).

ihhub commented:
Hi @Branikolog , could you please also help with testing of these changes are they are very critical for the release?

LeHerosInconnu commented:
Hello @ihhub,

Here are the differences I've spotted so far.
(One image from the original game with one image from fheroes2.)
There are many.
[Passability differences 01 01.zip](https://github.com/ihhub/fheroes2/files/6729549/Passability.differences.01.01.zip)
[Passability differences 01 02.zip](https://github.com/ihhub/fheroes2/files/6729560/Passability.differences.01.02.zip)
[Passability differences 01 03.zip](https://github.com/ihhub/fheroes2/files/6729577/Passability.differences.01.03.zip)

ihhub commented:
Hi @idshibanov  and @oleg-derevenetz , based on feedback from @LeHerosInconnu  and @Branikolog  I modified the logic. I also added temporary field called `_level` to avoid being overridden while we set artifact/spell/monster quantity. Could you please take a look at these changes again?

oleg-derevenetz commented on src/fheroes2/maps/maps_tiles.cpp:
```diff
@@ -2334,12 +2399,16 @@ void Maps::Tiles::setAsEmpty()
 {
-    msg >> ta.level >> ta.uniq >> ta.object >> ta.index >> ta.tmp;
+     uint8_t temp = 0;
+
```
I suppose, version check will be added later to not write/read an obsolete field.

ihhub replied:
Yes. I didn't want to increase this pull request even further.

oleg-derevenetz reviewed (approved):
Hi @ihhub no obvious issues found, except one note regarding read & write an obsolete field.

ihhub commented:
Hi @LeHerosInconnu and @Branikolog , could you please verify changes again? I addressed the majority of found problems. Not everything is identical to the original game but this implementation must be much closer to the original.

ihhub commented:
Hi @LeHerosInconnu and @Branikolog , for any other passability issues let's create a separate issue where we could define them. I found that the original game has some flaws in passability as well.
